Person Name Checks

CPBridge opened the pull request:
As described and discussed in #56, how to construct DICOM PersonNames correctly is not obvious and currently highdicom provides no guidance on how to do this correctly nor checks on strings provided as name parameters through the API. Part of the solution was to add a new constructor method in pydicom (see [pydicom#1331](https://github.com/pydicom/pydicom/pull/1331)) to simplify construction called `PersonName.from_named_components()`. This PR makes some changes on the highdicom side to close #56.

Specifically:
- A check for incorrect person names has been added in a new module `higidicom.valuerep`. This cannot be perfect, but will catch the very common case of people just supplying names with no formatting applied at all. If this check fails, it raises an exception with a useful error message that contains references about how to construct person names correctly and points to the new pydicom constructor (even though at the time or writing that change has still not been released in pydicom)
- Add calls to the check wherever a person name is passed to the highdicom API.
- Tweak the API such that wherever a parameter is passed that ultimately gets encoded as a PersonName, a `pydicom.value.PersonName` may be passed (previously this was allowed in some places but not others). This means that users can always use the new `PersonName` constructor.

hackermd reviewed (changes requested):
Thanks @CPBridge. Looks great. I think we should also apply this logic to SR templates and content items with value type PNAME ([highdicom.sr.value_types.PnameContentItem](https://github.com/MGHComputationalPathology/highdicom/blob/b378a20a23a62a55ef2614377465a034c42eb128/src/highdicom/sr/value_types.py#L208)). For example, `highdicom.sr.templates.PersonObserverIdentifyingAttributes`, which I just realized has the wrong value type (see #76).

CPBridge commented:
@hackermd I already added a check to the PnameContentItem constructor itself to catch all these cases, so I don't think any more is required? If you were seeing issues, it could be because of #76 . I will fix #76 in this PR shortly

CPBridge commented:
Fixed #76 by changing the value type of the person name in `PersonObserverIdentifyingAttribtutes` to `PnameContentItem`. Added some basic integration tests for `PnameContentItem` and `PersonObserverIdentifyingAttribtutes`.

hackermd commented on src/highdicom/sr/sop.py:
```diff
@@ -545,11 +547,11 @@ def __init__(
             SR document instance
-        verifying_observer_name: Union[str, None], optional
-            Name of the person that verfied the SR document
+        verifying_observer_name: Optional[Union[str, PersonName]], optional
```
```suggestion
        verifying_observer_name: Union[str, pydicom.valuerep.PersonName], optional
```

CPBridge replied:
Resolved in 3373a1f

hackermd commented on src/highdicom/sr/sop.py:
```diff
@@ -673,11 +675,11 @@ def __init__(
             SR document instance
-        verifying_observer_name: Union[str, None], optional
-            Name of the person that verfied the SR document
+        verifying_observer_name: Optional[Union[str, PersonName]], optional
```
```suggestion
        verifying_observer_name: Union[str, pydicom.valuerep.PersonName], optional
```

CPBridge replied:
Resolved in 3373a1f

CPBridge commented on src/highdicom/sr/sop.py:
```diff
@@ -417,11 +419,11 @@ def __init__(
             SR document instance
-        verifying_observer_name: Union[str, None], optional
-            Name of the person that verfied the SR document
+        verifying_observer_name: Optional[Union[str, PersonName]], optional
```
Thanks for spotting that `PersonName` should be fully qualified, but I don't see why we should get rid of `Optional`, as it is part of the type. Note that `Optional` and `optional` are not at all the same thing ([official docs](https://docs.python.org/3/library/typing.html#typing.Optional)) and in my opinion the docstrings should be as unambiguous as possible. 

If you prefer we could do `Union[str, pydicom.valuerep.PersonName, None]` rather than `Optional[Union[str, pydicom.valuerep.PersonName]]`? These are equivalent

@hackermd

CPBridge replied:
Now I look at it, the docstrings for many of the parameters of these methods are arguably incorrect...

hackermd replied:
I would prefer `Union[str, pydicom.valuerep.PersonName, None], optional` over `Optional[Union[str, pydicom.valuerep.PersonName]], optional` in the docsting (I would keep `Optional` for the type annotation), since I find the double optional confusing.

In general, I think using `Union[int, float], optional` rather than `Union[int, float, None], optional` is more intuitive (while maybe not being 100% semantically correct). In the majority of cases, we want to tell users that they can either pass an `int` or `float` object to the method and `None` is an artifact of making the parameter optional.
In this particular case, it's different, because the user may actually want to set the value of the corresponding type 2 DICOM attribute to `None` (empty value).

However, I agree with you that it may be better to be unambiguous and just annotate all possible types. We may then want to revisit the docstrings throughout the library and consider using a consistent type declaration in docstrings.

CPBridge replied:
Definitely agree that it's unfortunate that "Optional[...], optional" is technically the right thing to do here, but would argue that we are sort of stuck with it. If a user is dynamically determining the parameter value using some process that may results in `None`, it is important to be unambiguous that a value of `None` can be passed, in my opinion.

Anyway, for now I have switched to `Union[str, pydicom.valuerep.PersonName, None]` in 3373a1f , as suggested.

hackermd replied:
> Definitely agree that it's unfortunate that "Optional[...], optional" is technically the right thing to do here, but would argue that we are sort of stuck with it. If a user is dynamically determining the parameter value using some process that may results in None, it is important to be unambiguous that a value of None can be passed, in my opinion.

Agreed. Explicit is better than implicit :)

CPBridge commented:
I have updated the docstrings as agreed in the comments discussion
